File: 081. Search in Rotated Sorted Array II/solution-logn.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {boolean}
 */

// Time: O(logn), worst case O(n) when nearly all elements are the same
var search = function(nums, target) {
    if (nums === null || nums.length === 0) {
        return false;
    }
    
    var left = 0;
    var right = nums.length - 1;
    
    while (left <= right) {
        var mid = left + parseInt((right - left) / 2);
        
        if (nums[mid] === target) {
            return true;
        }
        
        if (nums[mid] > nums[left]) {
            if (nums[left] <= target && target < nums[mid]) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        } else if (nums[mid] < nums[left]) {
            if (nums[mid] < target && target <= nums[right]) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        } else {                            // where worst case might happen
            left++;
        }
    }
    
    return false;
};

File: Leet-081. Search in Rotated Sorted Array II/solution-logn.java
// Worst case is O(n), when nearly all elements are the same
class Solution {
    public boolean search(int[] nums, int target) {
        if (nums == null || nums.length == 0) {
            return false;
        }
        
        int left = 0, right = nums.length - 1;
        while (left <= right) {
            int mid = left + (right-  left) / 2;
            
            if (nums[mid] == target) {
                return true;
            }
            
            if (nums[mid] > nums[left]) { // left part is sorted
                if (nums[left] <= target && target < nums[mid]) {
                    right = mid - 1;
                } else {
                    left = mid + 1;
                }
            } else if (nums[mid] < nums[left]) {    // right part is sorted
                if (nums[mid] < target && target <= nums[right]) {
                    left = mid + 1;
                } else {
                    right = mid - 1;
                }
            } else {    // simply just update left
                left++;
            }
        }
        
        return false;
    }
}